Df1462 calendar icon

anselmbradford opened the pull request:
- Adds initial calendar icon CSS that looks like:

![screen shot 2015-02-17 at 3 40 02 pm](https://cloud.githubusercontent.com/assets/704760/6236759/4c3393a8-b6bb-11e4-8aaf-40546e8e9744.png)

- Removes excess whitespace from Gruntfile.

anselmbradford commented:
@Scotchester When you have a moment could you provide a review here. Still learning the code practices for the project so have some questions:

- Does the topdoc look correct? In particular I wasn't sure how to specify the family and tags and what the relationship (if any) there is between those two https://github.com/cfpb/cfgov-refresh/pull/254/files#diff-b442e0512c09298e8d797e46aad8afeeR6
- Where do component-specific variables go? https://github.com/cfpb/cfgov-refresh/pull/254/files#diff-b442e0512c09298e8d797e46aad8afeeR30
- The scale setting works great (change it to 2 and you get an icon twice as big), but this is probably not a practical approach for this. Better would be to be able to set the width or height and have the font-size, margin, padding, etc. cascade from that. Maybe best of all would be a font-size-like setting, so that this icon could be treated as if it were a typography icon like in cf-icons. Are there any existing examples of components that do this kind of scaling? https://github.com/cfpb/cfgov-refresh/pull/254/files#diff-b442e0512c09298e8d797e46aad8afeeR32
- I had to override the default line-height, which doesn't seem great. I could probably adjust the padding/margin so that it doesn't depend on overriding the default line-height, but at the same time, should the component set this value anyway so it's not dependent on global styles cascading down at all? https://github.com/cfpb/cfgov-refresh/pull/254/files#diff-b442e0512c09298e8d797e46aad8afeeR43
- Semantically `<time>` fits for this component, but there isn't native IE7 or IE8 support so there would need to be a polyfill I believe. Is it better to go semantically correct and patch it, or just use a `<div>` instead? I'd lean toward semantically correct myself. https://github.com/cfpb/cfgov-refresh/pull/254/files#diff-a755d1f813be365edd6b99bd615b6f87R25

jimmynotjim commented:
You could make this a mixin and pass in the background, foreground and scale eliminating the need to declare them ahead of time and allowing them to be more dynamic depending on use/location. This would allow you to solve a few of the issues you mentioned above. I put together [a quick CodePen](http://codepen.io/jimmynotjim/pen/e6bbbc1c0b188756c8d2d6ddcda30d9d) to explain. 

Using pseudo elems I was able to keep the full date as the content but visually hiding it and passing in the month and day as data-attrs for the `content` values. By using ems I was able eliminate a bit of the math involved and drop the height value. Width is unfortunately required but by setting it as `min-width` we can be sure it won't break if something oddly large gets used (the last examples show how this could happen).

anselmbradford commented:
@jimmynotjim This is looking great to me! Thanks for putting this together. One thought:
What about passing in the month/day value to the mixin as well, so have:
```less
.cf-calendar-icon(@month, @day, @background, @foreground, @scale) {
```

and 

```
.cf-calendar-icon(attr(data-month), attr(data-day), @gray-50, @white, 1);
```

That way the actual source of the month and date text doesn't need to be hardcoded to the `data-month` and `data-day` attribute names in the mixin, and the mixin could be used for other circumstances, like having the day and month swapped in the icon (if needed).

I read on chat from @Scotchester that `<time>` is good to use.

Scotchester commented:
**Topdoc** block looks good. I am pretty sure that `family` and `tags` are always the same; looks like you used the right values there.

You'll want to add `calendar-icon` to the `topdoc_families` array [in the Gruntfile](https://github.com/cfpb/cfgov-refresh/blob/gh-pages/Gruntfile.js#L387-L401) so that the docs page for this will be generated.

Also, now would be a good time to mention that [while our standards are going to be reviewed and potentially revised](https://github.com/cfpb/front-end/issues/), we do have some basic CSS standards in place. Largely, we follow [Idiomatic CSS](https://github.com/necolas/idiomatic-css) with a few suggested changes to the property order. Please see the Dev Manual, Coding Standards, CSS page.  Specifically on this code, I'd like to see updates to the property order, and the cfgov-refresh project tries to stick to 4-space indents in CSS/Less (and HTML).

(N.B.: We haven't yet determined a dev-team-wide indentation standard, so for now we just try to keep code within each project internally consistent.)

Scotchester commented:
**Component-specific variables**: That seems like a fine spot for them.

Scotchester commented:
*Scaling*: Seems like you're working this out with Jimmy's help.

Scotchester commented:
**Line height**: I think it is _completely_ appropriate to set `line-height: 1` in circumstances like this where you're trying to do a carefully-sized layout, and you won't ever have two lines of text next to each other.

Scotchester commented:
Waitaminute... I see that you HAVE generated the `/docs/calendar-icon/` page. How'd you do that without updating the `topdoc_families` array in the Gruntfile?

(Also, one benefit of using personal forks is that if you work on the `gh-pages` branch, you can preview docs updates on GH Pages. Not to say that outweighs your reasons for not doing that, but I was partly mistaken about this because I looked for http://anselmbradford.github.io/cfgov-refresh/docs/calendar-icon/ and nothing came up.)

jimmynotjim commented:
Entirely agree with adding the month/day as arguments. That will make the mixin more dynamic and easier to control in differing situations.

anselmbradford commented:
> Waitaminute... I see that you HAVE generated the /docs/calendar-icon/ page. How'd you do that without updating the topdoc_families array in the Gruntfile?

Hmm, I'm pretty sure this is what happened... I was confused at first how to include a new file in `main.min.css` before I discovered the imports in `main.less`. I thought at first you had to add the less file to `topdoc_families` in the Gruntfile, so I did that and ran `grunt`, but then noticed that `activity` wasn't included in that list, so I removed it again, went on with working on the project and forgot to add it back. But the docs still showed up from the initial time that I added it, so I'll need to re-add it there.

> ... cfgov-refresh project tries to stick to 4-space indents in CSS/Less (and HTML).

2 or 4 is fine with me. I lean toward 2-space myself because it fits better when working on smaller screen sizes, but makes sense to stick with project-wide standard. FWIW the [generated CSS file](https://github.com/cfpb/cfgov-refresh/blob/gh-pages/static/css/main.css) is 2-space and OAH appears to be 2-space. If it's easy to answer, otherwise I'll google it: can you set a per folder setting for spacing in Sublime, so when working in particular project any new files are already set to the project's spacing setting?


@jimmynotjim I think (pending the cleanup in the comments so far) this works for an MVP for this pattern for this sprint? I still think we need to consider how the scaling works, since I think it would be better if this worked consistent with how sizing works for an icon font, where you'd set a `font-size` for sizing. e.g.:

```less
.cf-calendar-icon(@month, @day, @background, @foreground, @font-size) {
```

Then behind the scenes the mixin would use the `@font-size` value to do the scaling (of course in the implementation this wouldn't literally be using `font-size`).

However, at this point we'll only have the icon at one size, so I think a multiplication scale is fine. Also, per a discussion I had w/ Virginia, we'll want to consider whether to move this to `cf-enhancements.less` at some point for an eventual integration into cf-framework, if that makes sense to do.

Scotchester commented:
Yeah, we don't really care about the generated CSS, except for perhaps how it looks when someone downloads the precompiled version of the framework. I wonder if we can change that option in the Less compiler.

Spoiler alert for our forthcoming standards discussions, but here are my main arguments for four: CSS isn't very wide to begin with, and in HTML, because every line, for the most part, begins with a `<`, it feels much more difficult to read two-space indentation, to me.

Yes, OAH's standard is 2-space, and yes, that's different from Flapjack's :)

I think one of the goals for the front-end discussion is to pick one and stick with it in the future, with possible gradual migration of existing projects.

Scotchester commented:
For Sublime, perhaps look into having it set at the Project level? I've never tried it, though.

An added complication is that we've been using two-space indents within the Topdoc blocks, which is annoying when they're in the same file as our four-spaced Less. :\

duelj commented:
This looks great! This may not be the right place for this but I was looking at the icon and wondered if there was a quick/easy way that the grey outer stroke could have a rounded inner corner instead of the square inner corner it has now. If this is not a simple solution, by all means ignore it. I just thought it was worth asking about, it's small but I think it feels more refined that way. 
![screen shot 2015-02-18 at 1 08 19 pm](https://cloud.githubusercontent.com/assets/6873734/6253463/96f48dca-b76f-11e4-9857-c2f36b519c1a.png)

anselmbradford commented:
@duelj Yep that's an easy one! I'm add it to the updates later today.

duelj commented:
RAD.

![selfie-4](http://i.imgur.com/HQvREin.gif)

anselmbradford commented:
Okay all, think I got everything. Some notes:

- Added @jimmynotjim 's mixin, with the following edits:
  - Added a border to the icon instead of padding/margin because of a pixel-rounding error with converting the ems to pixels https://github.com/cfpb/cfgov-refresh/pull/254/files#diff-b442e0512c09298e8d797e46aad8afeeR38
  - Changed mixin parameters `@background` and `@foreground` to `@background-color` and `@foreground-color`
  - Added `@month` and `@day` parameters.
- Added grouping of CSS properties. Are there other idiomatic styles I'm missing?
- Added `calendar-icon` to `topdoc_families` and regenerated output.
- Added `.less` suffix to less import statement.
- Removed example code snippet from index page.

Looks like (not actual size):
![screen shot 2015-02-18 at 5 06 51 pm](https://cloud.githubusercontent.com/assets/704760/6258776/79a45ae6-b799-11e4-8d36-5ac255ee773b.png)


If this looks good, give me :+1: and I'll squash in prep for merge.

jimmynotjim commented on static/css/calendar-icon.less:
```diff
@@ -0,0 +1,92 @@
+
+/*  Mixin to generate a calendar icon.
+    @month [String] The month (3-letter abbreviation, typically).
+    @day [Number] The day in the month.
```
Just had a thought, should probably use date and not day since day is often thought as the day of the week

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/3b149a0f5def00bf854148daeb61fe03a2e26f4d

jimmynotjim commented on static/css/calendar-icon.less:
```diff
@@ -0,0 +1,92 @@
+/*  Mixin to generate a calendar icon.
+    @month [String] The month (3-letter abbreviation, typically).
+    @day [Number] The day in the month.
+    @background-color [Color] A color value for the color behind the month.
```
Suggested change to "color value for the icon background" because it's not just the month

anselmbradford replied:
Changed to `@background-color [Color] A color value for the icon background (behind the month)`  in https://github.com/cfpb/cfgov-refresh/commit/ec6f6df230e17c72e05be017c8d5f1de3e34d7ac. I just wanted to include the month/day as references so if looking at the icon it would be easy to quickly tell which var was pointing to which color, if that's not already obvious.

jimmynotjim commented:
Go ahead and squash and I'll merge this in

anselmbradford commented:
:+1:
